runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -102,15 +102,23 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
       end_lnum = _end.line,
       end_col = line_byte_from_position(buf_lines, _end.line, _end.character, offset_encoding),
       severity = severity_lsp_to_vim(diagnostic.severity),
-      message = diagnostic.message
+      message = diagnostic.message,
+      user_data = {
+        lsp = {
+          code = diagnostic.code,
+          codeDescription = diagnostic.codeDescription,
+          tags = diagnostic.tags,
+          relatedInformation = diagnostic.relatedInformation,
+        },
```

Comment:
Could you also add the `data` property? 

```
 /**
      * A data entry field that is preserved between a
      * `textDocument/publishDiagnostics` notification and
      * `textDocument/codeAction` request.
      *
      * @since 3.16.0
      */
    data?: unknown;
```